rust/cadeau/src/xmf/vpx/mod.rs

```diff
@@ -0,0 +1,238 @@
+use core::fmt;
+
+use xmf_sys::{
+    XmfVpxCodecType, XmfVpxDecoderError, XmfVpxEncoderError, XmfVpxFrame, XmfVpxFrame_Destroy, XmfVpxFrame_GetBuffer,
+    XmfVpxFrame_GetDuration, XmfVpxFrame_GetFlags, XmfVpxFrame_GetHeight, XmfVpxFrame_GetPartitionId,
+    XmfVpxFrame_GetPts, XmfVpxFrame_GetSize, XmfVpxFrame_GetSpatialLayerEncoded, XmfVpxFrame_GetWidth, XmfVpxImage,
+    XmfVpxImage_Destroy, XmfVpxPacket, XmfVpxPacketKind, XmfVpxPacket_Destroy, XmfVpxPacket_GetFrame,
+    XmfVpxPacket_GetKind, XmfVpxPacket_IsEmpty,
+};
+
+pub mod decoder;
+pub mod encoder;
+
+#[derive(Debug, Clone, Copy)]
+pub enum VpxCodec {
+    VP8,
+    VP9,
+}
+
+pub struct VpxImage {
+    pub(crate) ptr: *mut XmfVpxImage,
+}
+
+impl Drop for VpxImage {
+    fn drop(&mut self) {
+        // Safety: it is safe to call, the pointer is owned by the VpxImage itself.
+        unsafe {
+            XmfVpxImage_Destroy(self.ptr);
+        }
+    }
+}
+
+impl From<VpxCodec> for XmfVpxCodecType {
+    fn from(codec: VpxCodec) -> Self {
+        match codec {
+            VpxCodec::VP8 => XmfVpxCodecType::VP8,
+            VpxCodec::VP9 => XmfVpxCodecType::VP9,
+        }
+    }
+}
+
+pub fn is_key_frame(buffer: &[u8]) -> bool {
+    if buffer.is_empty() {
+        return false;
+    }
+    buffer[0] & 0x1 == 0
+}
+
+/// Safety Note: The packet is only valid before the next call to any function on the encoder.
+/// The Packet iterator holds a mutable reference to the encoder,
+/// While the iterator is alive, the encoder with mutable reference is not allowed to be used.
+/// Hence all the functions on the packet are safe to call.
+///
+/// ex.
+///
+/// ```rust
+/// // This will not compile, as packet outlives the iterator.
+/// fn example(&mut self) {
+///     let mut packet;
+///     {
+///         let mut iterator = self.encoder.packet_iterator();
+///         packet = iterator.next().unwrap();
+///     }
+/// }
+/// ```
+/// // This will not compile as well, as the encoder is used after the iterator is created.
+/// ```rust
+/// fn example(&mut self) {
+///     let mut iterator = self.encoder.packet_iterator();
+///     let packet = iterator.next().unwrap();
+///     self.encoder.flush();
+/// }
+/// ```
+pub struct VpxPacket<'a> {
+    ptr: *mut XmfVpxPacket,
+    _marker: std::marker::PhantomData<&'a ()>,
+}
```

Comment:
question: Since we are freeing the memory in the drop implementation of this type, it means we are owning the resource. If, internally, XmfVpxPacket has a mutable pointer to VpxEncoder, then this should be made clear with:
```rust
_marker: std::marker::PhantomData<&'a mut VpxEncoder>,
```
However, I checked the C code, and XmfVpxPacket has no pointer to `VpxEncoder`. It also owns everything including the `vpx_codec_cx_pkt_t`.
As I understand, we don’t need any lifetime in VpxPacket itself.

This doc is only relevant for `PacketIterator` itself.

Reply:
I think it is important to have `VpxPacket` lifetime tied to `PacketIterator`, otherwise this could cause problem in the following example
```rust
let mut iterator = self.encoder.packet_iterator();
let packet = iterator.next().unwrap();
drop(iterator)
self.encoder.flush();
// packtet is still alive after this, but it is not guaranteed to be valid anymore.
```

XmfVpxPacket  does not own `vpx_codec_cx_pkt_t`, in fact, it is owned by `vpx_codec_ctx_t` inside `XmfVpxEncoder`, notice that `XmfVpxPacket_Destroy` function does not free `vpx_codec_cx_pkt_t`

Reply:
The packet is more like an reference from an internal buffer of `vpx_codec_cx_pkt_t` and libvpx may or may not modify the memories in that buffer during calls to `vpx_codec_ctx_t`, so we need to make sure the packet is only valid between calls of the encoder (i.e. `vpx_codec_ctx_t`) and limit the life time of the packet within the life time of the iterator would help to make this bound detectable by rust compiler.

Reply:
Oh, okay! Then you’re perfectly right here. Do you think you could add that to the doc then?

Also modify the marker to:

```rust
_marker: std::marker::PhantomData<&'a mut XmfVpxEncoder>
```

It’s best to use `&mut T` where `T` is the type on which we have a reference to conceptually to ensure we have the proper variance:
- https://doc.rust-lang.org/nomicon/subtyping.html
- https://doc.rust-lang.org/nomicon/phantom-data.html

Reply:
Oh, actually if you would consider the reference to the encoder as const/shared (we won't modify the encoder or its fields), you can go with this:

```rust
_marker: std::marker::PhantomData<&'a XmfVpxEncoder>
```

Without the `mut`. Only if you are sure the encoder is never modified while the VpxPacket is alive.

Reply:
updated!